### PR TITLE
Load images in runtime as buffers

### LINKED ISSUE #8329: Add option to return buffer from image.load
_Is it a private platform specific feature? (i.e. would it break your NDA?)<br>
Then please add your feature request in the corresponding extension repository!_

**Is your feature request related to a problem? Please describe (REQUIRED):**

**Describe the solution you'd like (REQUIRED):**

**Describe alternatives you've considered (REQUIRED):**

**Additional context (OPTIONAL):**

### PR DESCRIPTION
**Pull request by Jhonnyg**
The `image` script module now has a new function to return a buffer object. The new function is called `image.load_buffer` and can be used to dynamically create texture resources in runtime:

```lua
local img = image.load_buffer(compressed_data)
-- img.buffer now contains an actual buffer object that can be passed into resource.create_texture
local tparams = {
	width  = img.width,
	height = img.height,
	type   = resource.TEXTURE_TYPE_2D,
	format = resource.TEXTURE_FORMAT_RGBA
}
local my_texture_id = resource.create_texture("/my_custom_texture.texturec", tparams, img.buffer)
```

Fixes #8329 

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/scripts/script_image.cpp [hunk 1]
```diff
@@ -70,7 +72,15 @@ namespace dmScript
...
-    * @param [premult] [type:boolean] optional flag if alpha should be premultiplied. Defaults to `false`
+    * @param options [type:table] A table containing options about how to load the image. Supported entries:
+    *
+    * `premultiply_alpha`
+    * : [type:boolean] if true, premultiply the alpha channel
+    *
+    * `buffer`
+    * : [type:boolean] if true, the 'buffer' parameter of the returned table will be a buffer object (see [ref:buffer.create] for more info)
+    * that can be used for creating texture or atlas resources.
+    *
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/scripts/script_image.cpp:83]: What should we call this parameter?

Comment (britzl) [reply engine/gamesys/src/gamesys/scripts/script_image.cpp:83]: I think `buffer` is fine. Do we want more formats in the future? If that is the case then perhaps `format` is better, with values such as `bytes` (ie string) and `buffer`. But perhaps `format` implies image format more than return value type? Perhaps `type` instead?

File: engine/gamesys/src/gamesys/scripts/script_image.cpp [hunk 2]
```diff
@@ -93,6 +103,24 @@ namespace dmScript
...
+    * local imgurl = "http://www.site.com/image.png"
+    * http.request(imgurl, "GET", function(self, id, response)
+    *         local img = image.load(response.response, { buffer = true })
+    *         local tparams = {
+    *             width  = img.width,
+    *             height = img.height,
+    *             type   = resource.TEXTURE_TYPE_2D,
+    *             format = resource.TEXTURE_FORMAT_RGBA }
+    *
+    *         local my_texture_id = resource.create_texture("/my_custom_texture.texturec", tparams, img.buffer)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_image.cpp:119]: One can also imagine the api allowing a `string`, and we'll create a temporary buffer on the fly.
Which might be handy in the future if someone want to downkload a resource, and add it directlty into the resource system (without "loading" it) (e.g. we have a new api which we're using for the Rive extension, where we simply insert new precompiled resources)

File: engine/gamesys/src/gamesys/scripts/script_image.cpp [hunk 3]
```diff
@@ -102,8 +130,29 @@ namespace dmScript
...
+                lua_getfield(L, -1, "buffer");
+                buffer_output = lua_isnil(L, -1) ? false : lua_toboolean(L, -1);
+                lua_pop(L, 1);
+
+                lua_pop(L, 1);
+            }
+            else
+            {
+                // DEPRECATED
+                premult = lua_toboolean(L, 2);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_image.cpp:154]: Ideally, we should find these places in the code and log a dmLogWarningOnce message about deprecation and the suggested solution. MIght help with migration later on when we want to finally remove it.

File: engine/gamesys/src/gamesys/scripts/script_image.cpp (hunk 2 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/scripts/script_image.cpp:119]: Another api would be to create a buffer directly from a string?

File: engine/gamesys/src/gamesys/scripts/script_image.cpp [hunk 4]
```diff
@@ -12,23 +12,25 @@
 // CONDITIONS OF ANY KIND, either express or implied. See the License for the
 // specific language governing permissions and limitations under the License.
 
+#include <assert.h>
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_image.cpp:15]: Unused?
And if it _is_ used, we're in scripting domain, where lua_error would be preferred?

File: engine/gamesys/src/gamesys/scripts/script_image.cpp (hunk 2 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/scripts/script_image.cpp:119]: Was an `image.load_buffer()` also considered?

While adding an extra argument to the options table is likely the least invasive change we could make, and also saves code size. But I also think it's a slightly odd api.

And it sets a precedence that we can return different (unrelated) types from the same function, which isn't usually how we code in general.

**Review by JCash — DISMISSED** [review]
I tentatively approve, but I wonder if any other apis were considered?

**Pull request merged**